[Monitoring/Reporting Stats] Move hardcoded Reporting stats type collection out of monitoring plugin

tsullivan opened the pull request:
Closes https://github.com/elastic/kibana/issues/12504

- Registers the reporting collector in `plugins/reporting/index.js` instead of `plugins/monitoring/server/kibana_monitoring/collectors/get_reporting_collector.js`
- `KIBANA_REPORTING_TYPE` moved from the Monitoring plugin to the Reporting plugin

Depends on:

- [x] https://github.com/elastic/kibana/pull/18939
- [x] https://github.com/elastic/kibana/pull/18987
- [x] https://github.com/elastic/kibana/pull/19098
- [x] https://github.com/elastic/kibana/pull/19232

tsullivan commented on x-pack/plugins/monitoring/server/kibana_monitoring/init_kibana_monitoring.js:
```diff
@@ -21,7 +21,8 @@ export async function initKibanaMonitoring(kbnServer, server) {
 
-    startCollector(kbnServer, server, client);
+    const collector = startCollector(kbnServer, server, client);
+    server.expose('typeCollector', collector); // NOTE: consumers need to wait on Monitoring plugin status go green, and that this object was successfully exposed
```
The note is relevant to line 162 in x-pack/plugins/reporting/index.js

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2779/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2820/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2829/)

tsullivan commented:
I've removed review, added WIP label.

Reporting is not registering its collector in this code yet. It looks like the monitoring plugin needs to wait until the registration method is exposed before the plugin can go green. Since exposing the collector object that has the registration method becomes a first-class concern of Monitoring, I want to make Monitoring's green status held up by that, to resolve the race condition.

tsullivan commented:
Added depends on https://github.com/elastic/kibana/pull/18939

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2863/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2939/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3024/)

tsullivan commented on x-pack/plugins/reporting/index.js:
```diff
@@ -146,7 +147,8 @@ export const reporting = (kibana) => {
       logConfiguration(config, message => server.log(['reporting', 'debug'], message));
 
-      const xpackMainPlugin = server.plugins.xpack_main;
+      const { xpack_main: xpackMainPlugin, monitoring: monitoringPlugin } = server.plugins;
```
This needs testing on what happens when Monitoring is disabled but Reporting is enabled. cc @pickypg

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3236/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3653/)

Stacey-Gammon commented on x-pack/plugins/monitoring/server/kibana_monitoring/lib/get_collector_types_combiner.js:
```diff
@@ -9,8 +9,8 @@ import {
   KIBANA_USAGE_TYPE,
-  KIBANA_REPORTING_TYPE,
 } from '../../../common/constants';
+import { KIBANA_REPORTING_TYPE } from '../../../../reporting/common/constants';
```
Is there a step two planned to get rid of this and the abstraction leak below?  Or maybe you can get rid of it now?

tsullivan replied:
This module that combines collection types together into a bulk payload unfortunately does have to stay, but just for the short term. It's to support the behavior of Kibana Monitoring that sends kibana stats to the `monitoring/_bulk` endpoint in ES. Unfortunately, this module needs to have a lot of awareness of the usage collectors because in the current schema, usage is a field of `kibana_stats`. 

![image](https://user-images.githubusercontent.com/908371/40378299-fe156e16-5da7-11e8-8849-ed1142180098.png)

We will take the responsibility of sending stats to ES out of Kibana and have Metricbeats used for monitoring all the systems. Once we do that, we can remove `get_collector_types_combiner`

tsullivan replied:
@Stacey-Gammon https://github.com/elastic/kibana/blob/6.x/x-pack/plugins/xpack_main/server/routes/api/v1/kibana_stats.js has the same story as this module

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3812/)
